pkgs/development/python-modules/blis/default.nix

```diff
@@ -3,46 +3,54 @@
   buildPythonPackage,
   fetchFromGitHub,
   setuptools,
-  cython_0,
+  cython,
   hypothesis,
   numpy,
   pytestCheckHook,
   pythonOlder,
-  gitUpdater,
+  blis,
+  numpy_2,
 }:
 
 buildPythonPackage rec {
   pname = "blis";
-  version = "0.7.11";
+  version = "1.0.0";
   pyproject = true;
 
-  disabled = pythonOlder "3.7";
+  disabled = pythonOlder "3.9";
 
   src = fetchFromGitHub {
     owner = "explosion";
     repo = "cython-blis";
-    rev = "refs/tags/v${version}";
-    hash = "sha256-p8pzGZc5OiiGTvXULDgzsBC3jIhovTKUq3RtPnQ/+to=";
+    rev = "refs/tags/release-v${version}";
+    hash = "sha256-XS6h2c+8BJ9pAvIX8340C4vRZEBRmEZc6/6tH7ooqNU=";
   };
 
   postPatch = ''
+    # The commit pinning numpy to version 2 doesn't have any functional changes:
+    # https://github.com/explosion/cython-blis/pull/108
+    # BLIS should thus work with numpy and numpy_2.
+    substituteInPlace pyproject.toml setup.py \
+      --replace-fail "numpy>=2.0.0,<3.0.0" numpy
+
     # See https://github.com/numpy/numpy/issues/21079
     # has no functional difference as the name is only used in log output
     substituteInPlace blis/benchmark.py \
-      --replace 'numpy.__config__.blas_ilp64_opt_info["libraries"]' '["dummy"]'
+      --replace-fail 'numpy.__config__.blas_ilp64_opt_info["libraries"]' '["dummy"]'
   '';
 
   preCheck = ''
     # remove src module, so tests use the installed module instead
     rm -rf ./blis
   '';
 
-  nativeBuildInputs = [
+  build-system = [
     setuptools
-    cython_0
+    cython
+    numpy
   ];
 
-  propagatedBuildInputs = [ numpy ];
+  dependencies = [ numpy ];
```

Comment:
Isn't that a bit weird to put it in both places?

Reply:
That's what upstream specifies. IIRC numpy is imported in setup.py.

Reply:
> That's what upstream specifies. IIRC numpy is imported in setup.py.

Hmm if so, the only good reason to still put Numpy also in the native inputs is for cross compilation, which will not work on NixOS with the way upstream uses the native Numpy in their `setup.py`:

https://github.com/explosion/cython-blis/blob/b33c2ed9cf7ed51d999a926c4950dd8dec4d0286/setup.py#L86C39-L86C58

I think that for cross compilation you'd have to patch that line, and for non-cross compilation, you'd get the host's Numpy available anyway, so there is no need for numpy in the `build-system` list IMO.

Reply:
We plan to separate build dependencies and runtime dependencies in the future. https://github.com/NixOS/nixpkgs/issues/272178
So, we need to add numpy to both build-system and dependencies.

Reply:
> We plan to separate build dependencies and runtime dependencies in the future. #272178 So, we need to add numpy to both build-system and dependencies.

I'm not sure I agree with everything said there. Even if I had agreed, https://github.com/NixOS/nixpkgs/pull/272179 should have been finished before you'd apply that policy upon other packages.

Reply:
The order is reversed.
Without proper separation, we cannot start #272179.

Reply:
> you'd get the host's Numpy available anyway, so there is no need for numpy in the `build-system` list IMO.

There is because in the future `dependencies` might not be available at build time in order to reduce the number of rebuilds cause by updates, overrides, etc.

Reply:
> > you'd get the host's Numpy available anyway, so there is no need for numpy in the `build-system` list IMO.
> 
> There is because in the future `dependencies` might not be available at build time in order to reduce the number of rebuilds cause by updates, overrides, etc.

I support this motivation and this goal, but in this case, due to the goal of upstream's `import numpy` in `setup.py`, it is completely incorrect to add `numpy` to `build-system`, but rather you should patch (downstream or upstream) the file to support specifying manually the host machine's Numpy headers.

Reply:
I agree that cross doesn't work as is. But at least this continues to work if `dependencies ` are unavailable.

Reply:
> I agree that cross doesn't work as is. But at least this continues to work if `dependencies ` are unavailable.

OK I understand now. Still it would have been nice if you had put a TODO comment or something. Also the intentions of https://github.com/NixOS/nixpkgs/issues/272178 should be communicated in the documentation IMO, even though not everything is implemented yet.